conda-store-server/conda_store_server/action/install_lockfile.py

```diff
@@ -16,6 +18,32 @@ def action_install_lockfile(
     with lockfile_filename.open("w") as f:
         json.dump(conda_lock_spec, f)
 
+    # This essentially disables caching by having a dedicated cache per build.
+    # This is necessary because concurrent jobs override packages in the shared
+    # cache, which leads to failed builds.
+    #
+    # For example, packages A and B use the same dependency. Package A starts
+    # the download process and puts the dependency in the cache. Package B sees
+    # that the dependency is in the cache and tries to use it, but the
+    # dependency can be in an invalid state, such as being not fully-copied or
+    # downloaded, because the cache operations are not atomic, so the build of
+    # package B fails.
+    #
+    # This is a known problem in conda and I'm not aware of any solution besides
+    # just disabling the cache by having a per-build temporary cache directory.
```

Comment:
Another solution is to download the dependency to a temporary file (so that has different name from the dependency file) and when download is complete, rename the temporary file to the correct file name. In this way, the package B can see only the valid dependencies in the cache.

Reply:
@pearu I don't think it's practical to do here. conda-store is a thin wrapper around conda that provides a user-friendly web UI for package management. This would essentially require me to reimplement conda's caching. So it seems to me it's something that should be solved upstream, probably using the symlink mechanism, like in Nix, because it's an atomic operation on Linux. But there's also Windows to think about as well, which might require a different solution.

Reply:
@nkaretnikov yes, fixing this upstream makes sense.